### PR TITLE
Feature/create resource with labels

### LINKED ISSUE #60295: Allow setting labels when running kubectl create configmap/secret --from-file
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What you expected to happen**:
When issuing `kubectl create configmap --from-file=...`, I'd like to be able to set labels on the created map instead of having to run a separate.
Having the same ability when creating generic secrets would also be useful.

/sig cli

### PR DESCRIPTION
**Pull request by luigi-riefolo**
**What type of PR is this?**

/kind feature
/sig cli

**What this PR does / why we need it**:
Allow setting labels when running kubectl create.

Example:
`kubectl create configmap my-config --from-env-file=file --label=k1=v1,k2=v2`
`kubectl create configmap my-config --from-file=file --label=k1=v1,k2=v2`
`kubectl create configmap my-config --from-literal=key1=config1 -b=k1=v1,k2=v2`

**Which issue(s) this PR fixes**:

Fixes #60295

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
This PR introduces a new flag `--add-label` for the command `kubectl create configmap`
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
This PR makes the creation of resources that contain labels easier by using only one command, instead of:
- labelling the resource after the creation (2 commands)
- adding the labels manually in the manifest (2 commands)
```

### CODE REVIEW

**Comment by diist** [comment]
Could we make that `--add-label` flag work for all `kubectl create` commands as opposed to just `kubectl create configmap`?

**Comment by luigi-riefolo** [comment]
> Could we make that `--add-label` flag work for all `kubectl create` commands as opposed to just `kubectl create configmap`?

@diist  that's a good point. I guess it depends on the k8s members.

**Comment by luigi-riefolo** [comment]
@juanvallejo @dshulyak @seans3 any idea when this can be reviewed?

**Comment by luigi-riefolo** [comment]
@seans3 @dshulyak @juanvallejo any chance this can be reviewed?

**Comment by rikatz** [comment]
/ok-to-test

@luigi-riefolo it's hard for folks to watch all the PRs, so one recommendation would be to ping us/send this on #sig-cli at Kubernetes Slack :)

Will try to review this later and check what folks think about this
/cc
/assign @rikatz @dougsland

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go]: I believe a is correct here.

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:148]: My guess you need to add i18n.T("Add a labe as key...."

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/configmap.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/generate/versioned/configmap.go]: Please improve the docstring here "// addLabels" doesn't help too much.

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/configmap.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/generate/versioned/configmap.go]: Not sure what's the idea with these comments.

File: staging/src/k8s.io/kubectl/pkg/util/util.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/util/util.go]: Why not use directly ParseLiteralSource?

**Review by dougsland — CHANGES_REQUESTED** [review]
Hi @luigi-riefolo, thanks for PR. I agree with @rikatz, next time please nag us via slack. I see your PR needs a rebase. Do you mind to rebase it and squash the patches?

- Please also not use subject/description of patch: "up", "cosmetic" and such generic words.
- Looks like having such command in kubectl create would help users in different sub-commands.  Have you considered such approach?

**Comment by rikatz** [comment]
My points here:

I guess it would be better to have labels created for any object instead of only for config_map. As far as I know (Doug, correct me here please) any object on Kubernetes can be labeled, as this is part of metav1.ObjectMeta that every object have.

As an example, as a user I would like to have kubectl create namespace -l app=xpto -l env=prod myns (or ingress, or secret, etc etc!)

Maybe one approach here could be:

* Add on pkg/cmd/util/helpers.go a new function (as AddFieldManagerFlagVar) maybe called AddLabelFlagVar that receives the command and the pointer of a new array of strings (and the unit test!) and adds the flag --label/-l to the command. I'm suggesting to use ``StringSliceVarP`` here, so you can use syntax like -l label1=bla,label2=lala,label3=blo but it's up to other folks from sig-cli to accept this or not :)

* For each command that we want to support label when creating it, we should add the ``cmdutil.AddLabelFlagVar(cmd, o.Labels)``, and also change the command options to have a new Labels []string on the struct to receive that value

* We should have a common function (I guess there's already one, because of label on other parts of the code) that builds a map[string]string based on o.Labels array

* The last step is, anyone wanting to add those Labels to the object, should change the create command to check:
```
if len(labels) > 0 {
     obj.metadata.Labels = labels # Correct to the right path here, I'm just brainstorming and didn't checked the right syntax!
}
```

OK, this is my opinion how we can make a broader usage of this to any object :) I guess also annotations can be something desired on all create objects (create_ingress support this, but because a lot of functionality on ingress controllers are still changed with annotations) but we can discuss this further.

The approach for create with labels, sounds good?

@eddiezane @dougsland opinions? :)

**Comment by rikatz** [comment]
@luigi-riefolo any thoughts about what I've commented before, of having labels for all create commands? We've discussed this during sig-cli meeting, and the thinking is that it might be good to have this as a global flag for the create cmds :D

**Comment by luigi-riefolo** [comment]
@dougsland @rikatz thanks for your feedback guys! I'll change the PR ASAP

**Comment by rikatz** [comment]
@luigi-riefolo Thank you! Let me know if you need some help with this :D

**Review by rikatz — CHANGES_REQUESTED** [review]
As per previous conversations, let's turn this into a more global flag, all create commands are going to make a good usage of a --labels flag

**Comment by rikatz** [comment]
It seems there are some errors in code:

vendor/k8s.io/kubectl/pkg/generate/generate_test.go:308:16: undeclared name: ParseLabels (compile)

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:304]: I would say you are still developing and debugging, right? :D

**Comment by rikatz** [comment]
/hold 

This is a WIP :)

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go
Comment (luigi-riefolo) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:304]: yes @rikatz definitely debugging. I already got a new commit that removes those lines and adds unit testing.
I have a question, given that this PR is for configmaps but we agreed that this global flag can be used for other resources, should I create different PRs for each resource?

**Comment by luigi-riefolo** [comment]
@rikatz that's fine, I'll change the PR name then.

Can you please review it together with @dougsland, so that if you like the approach, I can do the same for the other resources.

> Hum, if you can do that for the other commands, I would do in this PR.

For other commands, do you mean other resources like namespaces, deployments, etc?

@dougsland I'll squash all my commits once this PR gets approved, if that's fine with you

File: staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:441]: You can use StringSlice here (or StringSliceVar), so for each combination of "key=value,key=value" it became a new item in the array

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:63]: By my comment below, this can be an array ([]string). But I don't know if we want to label things created directly by files (like kubectl create -l mykey=myvalue -f mypod.yaml)

File: staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:765]: Maybe we could turn the function I've used in create_ingress (https://github.com/kubernetes/kubernetes/blob/v1.21.0/staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go#L287-L296) into something more global.

In this case, the function receive a string and convert into a map (needs more validation tho, I've just found a bug -> https://github.com/kubernetes/kubernetes/pull/101377)

This way we can always use the []string

Comment (rikatz) [reply staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:765]: Or we can just use this function as I saw you've got it from the generators part :)

Anyway, I think using StringSlice is easier (and you cannot have ',' in labels anyway) but let me know what you think!

**Review by rikatz — COMMENTED** [review]
some minor comments from my part

File: staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go
Comment (lauchokyip) [inline staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go:24]: don't need the `v1` :)

Comment (luigi-riefolo) [reply staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go:24]: it seems a results of running go imports with go-vim, strange. Thanks for spotting it

File: staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:769]: I know this function has been migrated from generators but...

If we always expect a string to happen here, is there any reason not changing this function signature to receive a string, and remove the interface validation? This is going to be used gradually on functions, and to parse labels, so I would change this not to be interface and simplify things :)

File: staging/src/k8s.io/kubectl/pkg/util/util.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/util/util.go:110]: Where is this function being used? I couldn't find in the rest of the code

**Review by rikatz — COMMENTED** [review]
Hi @luigi-riefolo 

left some more other comments.

Overall I think this is on the right path :)

Thank you!!

**Comment by rikatz** [comment]
/remove-sig api-machinery apps auth cloud-provider cluster-lifecycle instrumentation network node release scalability scheduling
/remove-area test release-eng provider/gcp kubelet kubeadm e2e-test-framework code-generation cloudprovider apiserver
/remove-kind api-change

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:137]: OK, reading better I don't think we need to add this here, as when using kubectl create without subcommands we expect a file containing everything and labels are not being changed later.

So this can be removed from the main create command

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_clusterrole.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_clusterrole.go:193]: As this is a validation that might happen on the client side, and with the approach to normalize exit codes, I think that all the cmdutil.ParseLabels should be moved to "Validate" instead of "Run", as this might be a syntax validation that happens always on the client-side.

We can expect some more feedback about this, your approach here IMO is good, the change that I'm proposing actually will have more lines as it would be:

* Add a new ParsedLabels map to be used later on the object creation
* add the flag
* Inside the Validate, populate the ParsedLabels with "cmdutil.ParseLabels(c.Labels)" if this does not return an error
* Later in Run, do a clusterRole.Labels = ParsedLabels

I know it sounds strange, and this is why I want some more opinions. While this is more "complex" it will help in a future to separate errors from client-side from errors of server-side.

But I'm also ok with this approach (I don't want to gate keep on this) and rethink in this approach in the future when exit codes normalization became something :)

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_cronjob.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_cronjob.go:199]: What if instead of changing this function signature, you do something like:

```golang
labels, err := cmdutil.ParseLabel(o.Labels)
if err != nil {
     return err
}
if o.Client != nil {
  cronJob := o.createCronJob()
  cronjob.Labels = labels
[...]
cronJobBeta := o.createCronJobBeta()
cronJobBeta.Labels = labels
```

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_deployment.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_deployment.go:253]: OK, here we have a "problem": what happens if I specify something like "--label app=blo" while my app is called "bla"? This can generate inconsistencies, we should probably verify if k!="app" and then merge it into the map

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_deployment.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_deployment.go:186]: Same comment from above, instead of changing the "createDeployment" IMO the best approach here is to create it, and then merge the maps :)

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:249]: And same comment here :)

**Review by rikatz — CHANGES_REQUESTED** [review]
@luigi-riefolo thank you very much for working on this (and being patient with my long delay!) 

I've left some comments, overall:
* The approach looks good to me
* I think that on a long term, moving all the ParseLabels from Run() to Validate() will be better, because:
1) It will make it easier to separate return codes once the exit code normalization ships
2) Instead of changing every create* function to return an error, you can have an early return inside validate, and simply add those labels as part of the object when you reach the Run function knowing that this map was previously validated and sanitized (like in the create_deploy that you need to drop any attempt to change the "app" label).

I'll defer my comments to @eddiezane and @soltysh for a final review about my comments, but feel free to reach out here or in Slack to discuss this better!

**Comment by lauchokyip** [comment]
hey @luigi-riefolo  thanks for the PR. If I am not mistaken, I think in order for the PR to proceed it might be better to attend one of the sig-cli meeting to discuss with the members. @rikatz has already put it on the [meeting agenda](https://docs.google.com/document/d/1r0YElcXt6G5mOWxwZiXgGu_X6he3F--wKwg-9UBc29I/edit#heading=h.7hqjndol0olh)  but he has been busy with his work so I think it's better if you can attend the meeting and discuss with others to see every `create` command needs a `label`

**Comment by rikatz** [comment]
We already discussed, I just forgot to remove from the agenda :)

Em qui., 27 de mai. de 2021 às 23:44, Chok Yip Lau ***@***.***>
escreveu:

**Comment by lauchokyip** [comment]
@rikatz my apologies lol. Thanks for the clarification

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_rolebinding_test.go
Comment (lauchokyip) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_rolebinding_test.go:25]: don't need a `v1` here :)

Comment (luigi-riefolo) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_rolebinding_test.go:25]: thanks, it's still the go linter (maybe gofmt)

**Review by soltysh — CHANGES_REQUESTED** [review]
After re-considering my initial approval which I stated during SIG-CLI meeting back in April, I agree with @eddiezane that this addition isn't bringing too much of a value to kubectl. The original idea behind create commands is that they should be used as simple commands to get you familiar with k8s resources. For bigger customizations there are multiple options ranging from directly tweaking yaml files, through kustomize or chaining kubectl commands.

/close

**Pull request closed without merging**
